mono/metadata/threads.c

```diff
@@ -1434,19 +1445,30 @@ ves_icall_System_Threading_Thread_GetCurrentThread (MonoError *error)
 	return MONO_HANDLE_NEW (MonoThreadObject, mono_thread_current ());
 }
 
+static MonoInternalThread*
+thread_handle_to_internal_ptr (MonoThreadObjectHandle thread_handle)
+{
+	return MONO_HANDLE_GETVAL(thread_handle, internal_thread); // InternalThreads are always pinned.
+}
+
 MonoBoolean
-ves_icall_System_Threading_Thread_Thread_internal (MonoThread *this_obj,
-												   MonoObject *start)
+ves_icall_System_Threading_Thread_Thread_internal (MonoThreadObjectHandle thread_handle, MonoObjectHandle start_handle, MonoError *error)
 {
-	ERROR_DECL (error);
 	MonoInternalThread *internal;
 	gboolean res;
+	MonoThread *this_obj = MONO_HANDLE_RAW (thread_handle);
+	MonoObject *start = MONO_HANDLE_RAW (start_handle);
 
 	THREAD_DEBUG (g_message("%s: Trying to start a new thread: this (%p) start (%p)", __func__, this_obj, start));
 
-	if (!this_obj->internal_thread)
-		ves_icall_System_Threading_Thread_ConstructInternalThread (this_obj);
-	internal = this_obj->internal_thread;
+	internal = thread_handle_to_internal_ptr (thread_handle);
+
+	if (!internal) {
+		ves_icall_System_Threading_Thread_ConstructInternalThread (thread_handle, error);
+		mono_error_assert_ok (error);
```

Comment:
Why is this an assert ok?

Reply:
It appears before and after to assume success. I'd rather fail an assert than null dereference, perhaps -- there are arguments either way. The processor is always doing the checks for you, they do work, asserting redundantly all over the place is bloating. I can remove it.

Reply:
Removing "ves_icall" and the assert go together, since the underlying piece upon refactor never sets error, ok.